### ISSUE TITLE
FOREACH ignores context variables?

### ISSUE THREAD

**Issue by TimMensch** [FalkorDB/FalkorDB#699]
With this context:

```cypher
CREATE (:Post {name:"a post"}),(:Tag {name: "a"}), (:Tag {name: "b"}), (:Tag {name: "c"})
```

If you run this code, I would expect the post to have three tags attached to it:

```cypher
MATCH (t: Tag)
WITH COLLECT(t) as tags

MATCH (p:Post {name:"a post"})

WITH p, tags
FOREACH (tag IN tags | MERGE (p)-[:hasTag]->(tag))
return p, tags
```

But when I run this query after running the above:

```cypher
MATCH (p) WITH p OPTIONAL MATCH (p)-[q]->(r) RETURN p,q,r
```

I get this:
<img width="814" alt="firefox_2024-06-06_09-01-13" src="https://github.com/FalkorDB/FalkorDB/assets/530873/284be70c-b7cc-4f0c-a28a-a3627f318892">

The "Post" has nothing attached to it, and three new, unlabeled nodes are created, one per tag.

I've seen examples in the Neo4J universe that clearly reference context variables from the surrounding context, so this feels like a bug.

**Comment by swilly22** [comment]
Hi @TimMensch you're absolutely right, this is a bug. thank you for reporting!
in the meanwhile let me suggest switching from the `FOREACH` clause to `UNWIND`
```cypher
MATCH (t: Tag)
WITH COLLECT(t) as tags

MATCH (p:Post {name:"a post"})

WITH p, tags

UNWIND tags as tag
MERGE (p)-[:hasTag]->(tag)

RETURN p, tags
```

**Comment by TimMensch** [comment]
Yeah, I started out with UNWIND. When I use UNWIND, I don't get a return value. Here's more of what I'm doing:

```cypher
UNWIND [] AS aTag
MATCH (t:Tag {tag: aTag})
WITH collect(t) as tags

MERGE (p: Post {name: "the post"})
ON CREATE SET
  p.createdAt=1234

WITH p,tags
UNWIND tags as tag
MERGE (p)-[:hasTag]->(tag)

RETURN p, tags
```

See the `UNWIND []`? That's what happens when a particular Post I'm creating doesn't have tags. 

When there are tags, it works. I get p and an array of the tag objects.

When there are no tags, I get `null` as a return from the query.

I've tried everything I could think of. `OPTIONAL MATCH` was the first thing I tried. No use. I tried isolating the code in a `CALL`. Nope. 

`FOREACH` was a last-ditch effort to _really really_ convince it to give me a return result even if part of the MATCH never happened.

I don't know whether the above behavior is a "bug", per se, but it's a poor behavior from my POV. The `FOREACH` was undeniably a bug, though, which is why I posted it.
